src/communication/communication_channel.cpp

```diff
@@ -0,0 +1,164 @@
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+/*
+ * communication_channel.cpp - wrapper for communication primitives
+ *                           - it can establish a connection, send/receive messages, wait for events and close the connection
+ */
+
+#include "communication_channel.hpp"
+
+#include "connection_support.h"
+#include "connection_globals.h"
+#include "system_parameter.h"
+#include "thread_manager.hpp"
+#include <string>
+#include "connection_sr.h"
+
+#if defined(WINDOWS)
+#include "wintcp.h"
+#else /* WINDOWS */
+#include "tcp.h"
+#endif /* WINDOWS */
+
+communication_channel::communication_channel (int max_timeout_in_ms) :
+  m_max_timeout_in_ms (max_timeout_in_ms),
+  m_type (CHANNEL_TYPE::NO_TYPE),
+  m_socket (INVALID_SOCKET)
+{
+}
+
+communication_channel::communication_channel (communication_channel &&comm) : m_max_timeout_in_ms (
+	  comm.m_max_timeout_in_ms)
+{
+  m_type = comm.m_type;
+  comm.m_type = NO_TYPE;
+
+  m_socket = comm.m_socket;
+  comm.m_socket = INVALID_SOCKET;
+}
+
+communication_channel &communication_channel::operator= (communication_channel &&comm)
+{
+  assert (!is_connection_alive ());
+  this->~communication_channel ();
+
+  new (this) communication_channel (std::forward <communication_channel> (comm));
+  return *this;
+}
+
+communication_channel::~communication_channel ()
+{
+  close_connection ();
+}
+
+int communication_channel::send (const std::string &message)
+{
+  return communication_channel::send (message.c_str (), message.length ());
+}
+
+int communication_channel::recv (char *buffer, int &received_length)
+{
+  return css_net_recv (m_socket, buffer, &received_length, m_max_timeout_in_ms);
+}
+
+int communication_channel::send (const char *buffer, int length)
+{
+  int templen, vector_length = 2;
+  int total_len = 0;
+  struct iovec iov[2];
+
+  css_set_io_vector (&iov[0], &iov[1], buffer, length, &templen);
+  total_len = sizeof (int) + length;
+
+  return css_send_io_vector_with_socket (m_socket, iov, total_len, vector_length, m_max_timeout_in_ms);
+}
+
+int communication_channel::connect (const char *hostname, int port)
+{
+  if (is_connection_alive ())
+    {
+      assert (false);
+      return ER_FAILED;
+    }
+
+  m_type = CHANNEL_TYPE::INITIATOR;
+  m_socket = css_tcp_client_open (hostname, port);
+
+  return IS_INVALID_SOCKET (m_socket) ? REQUEST_REFUSED : NO_ERRORS;
+}
+
+int communication_channel::accept (SOCKET socket)
+{
+  if (is_connection_alive () || IS_INVALID_SOCKET (socket))
+    {
+      return ER_FAILED;
+    }
+
+  m_type = CHANNEL_TYPE::LISTENER;
+  m_socket = socket;
+
+  return NO_ERROR;
+}
+
+void communication_channel::close_connection ()
+{
+  if (!IS_INVALID_SOCKET (m_socket))
+    {
+      css_shutdown_socket (m_socket);
+      m_socket = INVALID_SOCKET;
+      m_type = NO_TYPE;
+    }
+}
+
+const int &communication_channel::get_max_timeout_in_ms ()
```

Comment:
Is there a reason to return ```const int &``` instead of just ```int```?

Reply:
not really